File: objectModel/Python/cdm/resolvedmodel/resolved_attribute.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Any, cast, Optional, Union, TYPE_CHECKING

from cdm.resolvedmodel.resolved_trait_set import ResolvedTraitSet
from cdm.resolvedmodel.resolved_attribute_set import ResolvedAttributeSet

if TYPE_CHECKING:
    from cdm.objectmodel import CdmAttribute, CdmAttributeContext, CdmObject, SpewCatcher
    from cdm.resolvedmodel import AttributeResolutionContext, ResolvedAttributeSet
    from cdm.utilities import ApplierState, ResolveOptions, TraitToPropertyMap

    ResolutionTarget = Union[CdmAttribute, ResolvedAttributeSet]


class ResolvedAttribute():
    def __init__(self, res_opt: 'ResolveOptions', target: 'ResolutionTarget', default_name: str, att_ctx: 'CdmAttributeContext') -> None:
        self.applier_state = None  # type: Optional[ApplierState]
        self.arc = None  # type: Optional[AttributeResolutionContext]
        self.att_ctx = att_ctx  # type: CdmAttributeContext
        self.insert_order = 0  # type: int
        self.previous_resolved_name = default_name  # type: str
        self.resolved_traits = ResolvedTraitSet(res_opt)  # type: ResolvedTraitSet

        # Internal
        self._resolved_attribute_count = 0  # type: int
        self.target = target  # type: ResolutionTarget
        self._resolved_name = default_name  # type: str
        self._ttpm = None  # type: Optional[TraitToPropertyMap]
        # if the target is a resolved attribute set, then we are wrapping it. update the lineage of this new ra to point at all members of the set
        if isinstance(target, ResolvedAttributeSet) and att_ctx is not None:
            ras_sub = target  # type: ResolvedAttributeSet
            if ras_sub._set is not None and len(ras_sub._set) > 0:
                for ra_sub in ras_sub._set:
                    if ra_sub.att_ctx is not None:
                        att_ctx._add_lineage(ra_sub.att_ctx)

    @property
    def resolved_name(self) -> str:
        return self._resolved_name

    @resolved_name.setter
    def resolved_name(self, value: str) -> None:
        self._resolved_name = value
        if self.previous_resolved_name is None:
            self.previous_resolved_name = value

    @property
    def target(self) -> 'ResolutionTarget':
        return self._target

    @target.setter
    def target(self, value: 'ResolutionTarget') -> None:
        from cdm.objectmodel import CdmAttribute
        from cdm.resolvedmodel import ResolvedAttributeSet
        if value is not None:
            if isinstance(value, CdmAttribute) and value._attribute_count:
                self._resolved_attribute_count = value._attribute_count
            elif isinstance(value, ResolvedAttributeSet):
                self._resolved_attribute_count = value._resolved_attribute_count
        self._target = value

    @property
    def is_primary_key(self) -> Optional[bool]:
        return self._trait_to_property_map._fetch_property_value('isPrimaryKey')

    @property
    def is_read_only(self) -> Optional[bool]:
        return self._trait_to_property_map._fetch_property_value('isReadOnly')

    @property
    def is_nullable(self) -> Optional[bool]:
        return self._trait_to_property_map._fetch_property_value('isNullable')

    @property
    def data_format(self) -> str:
        return self._trait_to_property_map._fetch_property_value('dataFormat')

    @property
    def source_name(self) -> str:
        return self._trait_to_property_map._fetch_property_value('sourceName')

    @property
    def source_ordering(self) -> Optional[int]:
        return self._trait_to_property_map._fetch_property_value('sourceOrdering')

    @property
    def display_name(self) -> str:
        return self._trait_to_property_map._fetch_property_value('displayName')

    @property
    def description(self) -> str:
        return self._trait_to_property_map._fetch_property_value('description')

    @property
    def maximum_value(self) -> str:
        return self._trait_to_property_map._fetch_property_value('maximumValue')

    @property
    def minimum_value(self) -> str:
        return self._trait_to_property_map._fetch_property_value('minimumValue')

    @property
    def maximum_length(self) -> Optional[int]:
        return self._trait_to_property_map._fetch_property_value('maximumLength')

    @property
    def value_constrained_to_list(self) -> Optional[bool]:
        return self._trait_to_property_map._fetch_property_value('valueConstrainedToList')

    @property
    def default_value(self) -> Any:
        return self._trait_to_property_map._fetch_property_value('defaultValue')

    @property
    def creation_sequence(self) -> int:
        return self.insert_order

    @property
    def _trait_to_property_map(self) -> 'TraitToPropertyMap':
        from cdm.utilities import TraitToPropertyMap

        if self._ttpm is not None:
            return self._ttpm

        self._ttpm = TraitToPropertyMap(cast('CdmObject', self.target))
        return self._ttpm

    def copy(self) -> 'ResolvedAttribute':
        from cdm.resolvedmodel import ResolvedAttributeSet

        # Use the options from the traits.
        copy = ResolvedAttribute(self.resolved_traits.res_opt, self.target, self._resolved_name, None)
        copy.previous_resolved_name = self.previous_resolved_name
        copy.resolved_name = self.resolved_name
        copy._resolved_attribute_count = self._resolved_attribute_count
        copy.resolved_traits = self.resolved_traits.shallow_copy()
        copy.insert_order = self.insert_order
        copy.arc = self.arc
        copy.att_ctx = self.att_ctx  # set here instead of constructor to avoid setting lineage for this copy

        # deep copy when set contains sets. this copies the resolved att set and the context, etc.
        copy.target = copy.target.copy()

        from cdm.objectmodel import CdmAttribute
        if isinstance(copy.target, CdmAttribute):
            copy.target.owner = self.target.owner
            copy.target.in_document = self.target.in_document

        if self.applier_state is not None:
            copy.applier_state = self.applier_state._copy()

        return copy

    def spew(self, res_opt: 'ResolveOptions', to: 'SpewCatcher', indent: str, name_sort: bool) -> None:
        to.spew_line('{}[{}]'.format(indent, self._resolved_name))
        self.resolved_traits.spew(res_opt, to, indent + '-', name_sort)

    def complete_context(self, res_opt: 'ResolveOptions') -> None:
        from cdm.objectmodel import CdmAttribute

        if self.att_ctx:
            if self.att_ctx.name is None:
                self.att_ctx.name = self._resolved_name
                self.att_ctx.at_corpus_path = str(self.att_ctx.parent.fetch_object_definition(res_opt).at_corpus_path) + '/' + self._resolved_name

            if self.att_ctx.definition is None and isinstance(self.target, CdmAttribute):
                self.att_ctx.definition = self.target.create_simple_reference(res_opt)

File: objectModel/TypeScript/ResolvedModel/ResolvedAttribute.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    AttributeResolutionContext,
    CdmAttribute,
    CdmAttributeContext,
    CdmEntityDefinition,
    CdmObject,
    ResolvedAttributeSet,
    ResolvedTraitSet,
    resolveOptions,
    spewCatcher,
    traitToPropertyMap
} from '../internal';

/**
     * @internal
     */
export class ResolvedAttribute {

    public get isPrimaryKey(): boolean {
        return this.traitToPropertyMap
            .fetchPropertyValue('isPrimaryKey');
    }
    public get isReadOnly(): boolean {
        return this.traitToPropertyMap
            .fetchPropertyValue('isReadOnly');
    }
    public get isNullable(): boolean {
        return this.traitToPropertyMap
            .fetchPropertyValue('isNullable');
    }
    public get dataFormat(): string {
        return this.traitToPropertyMap
            .fetchPropertyValue('dataFormat');
    }
    public get sourceName(): string {
        return this.traitToPropertyMap
            .fetchPropertyValue('sourceName');
    }
    public get sourceOrdering(): number {
        return this.traitToPropertyMap
            .fetchPropertyValue('sourceOrdering');
    }
    public get displayName(): string {
        return this.traitToPropertyMap
            .fetchPropertyValue('displayName');
    }
    public get description(): string {
        return this.traitToPropertyMap
            .fetchPropertyValue('description');
    }
    public get maximumValue(): string {
        return this.traitToPropertyMap
            .fetchPropertyValue('maximumValue');
    }
    public get minimumValue(): string {
        return this.traitToPropertyMap
            .fetchPropertyValue('minimumValue');
    }
    public get maximumLength(): number {
        return this.traitToPropertyMap
            .fetchPropertyValue('maximumLength');
    }
    public get valueConstrainedToList(): boolean {
        return this.traitToPropertyMap
            .fetchPropertyValue('valueConstrainedToList');
    }
    public get defaultValue(): any {
        return this.traitToPropertyMap
            .fetchPropertyValue('defaultValue');
    }
    public get creationSequence(): number {
        return this.insertOrder;
    }
    private get traitToPropertyMap(): traitToPropertyMap {
        if (this.t2pm) {
            return this.t2pm;
        }
        this.t2pm = new traitToPropertyMap(this.target as CdmObject);

        return this.t2pm;
    }
    public previousResolvedName: string;
    /**
     * @internal
     */
    public resolvedTraits: ResolvedTraitSet;
    /**
     * @internal
     */
    public resolvedAttributeCount: number;
    /**
     * @internal
     */
    public owner: CdmEntityDefinition;
    public insertOrder: number = 0;
    public attCtx: CdmAttributeContext;
    public applierState?: any;
    public arc: AttributeResolutionContext;
    private _target: ResolutionTarget;
    private _resolvedName: string;
    private t2pm: traitToPropertyMap;

    constructor(resOpt: resolveOptions, target: ResolutionTarget, defaultName: string, attCtx: CdmAttributeContext) {
        // let bodyCode = () =>
        {
            this.target = target;
            this.resolvedTraits = new ResolvedTraitSet(resOpt);
            this.resolvedName = defaultName;
            this.previousResolvedName = defaultName;
            this.attCtx = attCtx;
            // if the target is a resolved attribute set, then we are wrapping it. update the lineage of this new ra to point at all members of the set
            if (target instanceof ResolvedAttributeSet && attCtx) {
                const rasSub: ResolvedAttributeSet = target as ResolvedAttributeSet;
                if (rasSub.set && rasSub.set.length > 0) {
                    for (const raSub of rasSub.set) {
                        if (raSub.attCtx) {
                            attCtx.addLineage(raSub.attCtx);
                        }
                    }
                }
            }
        }
        // return p.measure(bodyCode);
    }
    public set resolvedName(newVal: string) {
        this._resolvedName = newVal;
        if (this.previousResolvedName === undefined) {
            this.previousResolvedName = newVal;
        }
    }
    public get resolvedName(): string {
        return this._resolvedName;
    }
    public set target(newVal: ResolutionTarget) {
        if (newVal !== undefined) {
            if (newVal instanceof CdmAttribute) {
                this.resolvedAttributeCount = newVal.attributeCount;
            } else if (newVal instanceof ResolvedAttributeSet) {
                this.resolvedAttributeCount = newVal.resolvedAttributeCount;
            }
        }
        this._target = newVal;
    }
    public get target(): ResolutionTarget {
        return this._target;
    }
    public copy(): ResolvedAttribute {
        // let bodyCode = () =>
        {
            const resOpt: resolveOptions = this.resolvedTraits.resOpt; // use the options from the traits
            const copy: ResolvedAttribute = new ResolvedAttribute(resOpt, this.target, this._resolvedName, undefined);
            copy.previousResolvedName = this.previousResolvedName;
            copy.resolvedName = this.resolvedName;
            copy.resolvedTraits = this.resolvedTraits.shallowCopy();
            copy.insertOrder = this.insertOrder;
            copy.arc = this.arc;
            copy.attCtx = this.attCtx // set here instead of constructor to avoid setting lineage for this copy
            copy.owner = this.owner;

            if (copy.target instanceof ResolvedAttributeSet) {
                // deep copy when set contains sets. this copies the resolved att set and the context, etc.
                copy.target = copy.target.copy() as ResolvedAttributeSet;
            }

            if (this.applierState) {
                copy.applierState = {};
                Object.assign(copy.applierState, this.applierState);
            }

            return copy;
        }
        // return p.measure(bodyCode);
    }
    public spew(resOpt: resolveOptions, to: spewCatcher, indent: string, nameSort: boolean): void {
        // let bodyCode = () =>
        {
            to.spewLine(`${indent}[${this._resolvedName}]`);
            this.resolvedTraits.spew(resOpt, to, `${indent}-`, nameSort);
        }
        // return p.measure(bodyCode);
    }

    public completeContext(resOpt: resolveOptions): void {
        if (this.attCtx) {
            if (this.attCtx.name == null) {
                this.attCtx.name = this._resolvedName;
                this.attCtx.atCorpusPath = `${this.attCtx.parent.fetchObjectDefinition<CdmAttributeContext>(resOpt).atCorpusPath}/${this._resolvedName}`;
            }

            if (!this.attCtx.definition && this.target instanceof CdmAttribute)
            this.attCtx.definition = (this.target as CdmAttribute).createPortableReference(resOpt);
        }
    }
}

export type ResolutionTarget = (CdmAttribute | ResolvedAttributeSet | string);
